tests/test_plugin.py

```diff
@@ -2593,22 +2561,33 @@ def init(options, configuration, plugin):
 
 
 def test_plugin_shutdown(node_factory):
-    """test 'shutdown' notification"""
+    """test 'shutdown' notifications, via `plugin stop` or via `stop`"""
+
     p = os.path.join(os.getcwd(), "tests/plugins/test_libplugin")
-    l1 = node_factory.get_node(options={'plugin': p})
+    p2 = os.path.join(os.getcwd(), 'tests/plugins/misc_notifications.py')
+    l1 = node_factory.get_node(options={'plugin': [p, p2]})
 
     l1.rpc.plugin_stop(p)
     l1.daemon.wait_for_log(r"test_libplugin: shutdown called")
     # FIXME: clean this up!
     l1.daemon.wait_for_log(r"test_libplugin: Killing plugin: exited during normal operation")
 
-    # Now try timeout.
+    # Via `plugin stop` it can make RPC calls before it (self-)terminates
+    l1.rpc.plugin_stop(p2)
+    l1.daemon.wait_for_log(r'misc_notifications.py: via plugin stop, datastore success')
+    l1.rpc.plugin_start(p2)
+
+    # Now try timeout via `plugin stop`
     l1.rpc.plugin_start(p, dont_shutdown=True)
     l1.rpc.plugin_stop(p)
     l1.daemon.wait_for_log(r"test_libplugin: shutdown called")
     l1.daemon.wait_for_log(r"test_libplugin: Timeout on shutdown: killing anyway")
 
-    # Now, should also shutdown on finish.
-    l1.rpc.plugin_start(p)
+    # Now, should also shutdown or timeout on finish, RPC calls then fail with error code -5
+    l1.rpc.plugin_start(p, dont_shutdown=True)
+    needle = l1.daemon.logsearch_start
     l1.rpc.stop()
     l1.daemon.wait_for_log(r"test_libplugin: shutdown called")
+    l1.daemon.logsearch_start = needle          # we don't know what comes first
+    l1.daemon.is_in_log(r'misc_notifications.py: via lightningd shutdown, datastore failed')
+    l1.daemon.is_in_log(r'test_libplugin: failed to self-terminate in time, killing.')
```

Comment:
We have wait_for_logs() which might work better here?

Reply:
Thanks, that works and I learned (again) about `wait_for_logs`. Added to my TODO list.